src/mongoose_rabbit_worker.erl

```diff
@@ -0,0 +1,302 @@
+%%% Metrics
+%%%===================================================================
+
+-define(BACKEND, mod_event_pusher_rabbit).
+-define(CONNECTIONS_ACTIVE_METRIC, [backends, ?BACKEND, connections_active]).
+-define(CONNECTIONS_OPENED_METRIC, [backends, ?BACKEND, connections_opened]).
+-define(CONNECTIONS_CLOSED_METRIC, [backends, ?BACKEND, connections_closed]).
+-define(CONNECTIONS_FAILED_METRIC, [backends, ?BACKEND, connections_failed]).
+-define(MESSAGES_PUBLISHED_METRIC, [backends, ?BACKEND, messages_published]).
+-define(MESSAGES_FAILED_METRIC, [backends, ?BACKEND, messages_failed]).
+-define(MESSAGES_TIMEOUT_METRIC, [backends, ?BACKEND, messages_timeout]).
+-define(MESSAGE_PUBLISH_TIME_METRIC,
+        [backends, ?BACKEND, message_publish_time]).
+-define(MESSAGE_PAYLOAD_SIZE_METRIC,
+        [backends, ?BACKEND, message_payload_size]).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the RabbitMQ worker.
+%% @end
+%%--------------------------------------------------------------------
+start_link() ->
+    gen_server:start_link(?MODULE, [], []).
+
+list_metrics() ->
+    [{?CONNECTIONS_ACTIVE_METRIC, spiral},
+     {?CONNECTIONS_OPENED_METRIC, spiral},
+     {?CONNECTIONS_CLOSED_METRIC, spiral},
+     {?CONNECTIONS_FAILED_METRIC, spiral},
+     {?MESSAGES_PUBLISHED_METRIC, spiral},
+     {?MESSAGES_FAILED_METRIC, spiral},
+     {?MESSAGES_TIMEOUT_METRIC, spiral},
+     {?MESSAGE_PUBLISH_TIME_METRIC, histogram},
+     {?MESSAGE_PAYLOAD_SIZE_METRIC, histogram}].
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init(Opts) ->
+    process_flag(trap_exit, true),
+    self() ! {init, Opts},
+    {ok, #state{}}.
+
+handle_call(Req, From, State) ->
+    maybe_handle_request(fun do_handle_call/3, [Req, From, State],
+                         ?REPLY_REQ_DROPPED(State)).
+
+handle_cast(Req, State) ->
+   maybe_handle_request(fun do_handle_cast/2, [Req, State],
+                        ?NOREPLY_REQ_DROPPED(State)).
+
+handle_info(Req = {init, _}, State) ->
+    do_handle_info(Req, State);
+handle_info(Req, State) ->
+    maybe_handle_request(fun do_handle_info/2, [Req, State],
+                         ?NOREPLY_REQ_DROPPED(State)).
+
+terminate(_Reason, #state{connection = Connection, channel = Channel,
+                          host = Host}) ->
+    close_rabbit_connection(Connection, Channel, Host),
+    ok.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+do_handle_call({amqp_call, Method}, _From, State = #state{channel = Channel}) ->
+    case try_to_apply(amqp_channel, call, [Channel, Method]) of
```

Comment:
Use `try ... catch` here.
Dialyzer can't handle`erlang:apply/3` calls. `amqp_channel:call(Channel, Method)` should be used instead.

Reply:
I second @arcusfelis's remark. I do not see a reason not to use `amqp_channel:call(Channel, Method)` directly and also to hide the `try ... catch` block in a separate function. The overall structure of code is almost identical, while the cognitive overhead of non-local exception handling is avoided:

```erlang
try
    Res = ampq_channel:call(Channel, Method),
    {reply, Res, State}
catch E:R ->
    {FreshConn, FreshChann} = maybe_restart_rabbit_connection(State),
    {reply, {E, R}, State#state{connection = FreshConn,
                                channel = FreshChann}}
end
```